src/lib/Imp.hs

```diff
@@ -832,6 +838,8 @@ splitDest (maybeDest, (Block decls ans)) = do
       (_, Con (Lit _)) -> tell [(dest, result)]
       -- This is conservative, in case the type is dependent. We could do better.
       (DataConRef _ _ _, DataCon _ _ _ _) -> tell [(dest, result)]
+      -- This is conservative. Without it, we hit bugs like #348
+      (Con (ConRef (SumAsProd _ _ _)), Con (SumAsProd _ _ _)) -> tell [(dest, result)]
```

Comment:
Yes, it's necessary because we don't want the unused arms of the sum type to be treated as valid dests in their own right, because then they couldn't be skipped based on the tag. Good point about de-duping.